### PR TITLE
cli: add COCKROACH_MEMPROF_INTERVAL

### PR DESCRIPTION
**Pull request by petermattis**
The COCKROACH_MEMPROF_INTERVAL env var contains a duration which
indicates how often memory profiles will be generated. Generated
profiles are written to the logs directory with the prefix "memprof."
and a date/time suffix. Periodic memory profiling is only initiated for
the `start` command.

See #7018.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7027)

### CODE REVIEW

File: cli/start.go [hunk 1]
```diff
@@ -121,6 +124,42 @@ func initInsecure() error {
...
+		t := time.NewTicker(memProfileInterval)
+		for {
+			<-t.C
+
+			func() {
+				const format = "2006-01-02T15_04_05.999"
+				path := filepath.Join(dir, "memprof."+timeutil.Now().Format(format))
+				f, err := os.Create(path)
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "testing: %s\n", err)
```
Comment (tamird) [inline cli/start.go]: Did you mean to keep this?

File: cli/start.go [hunk 2]
```diff
@@ -121,6 +124,42 @@ func initInsecure() error {
...
+					os.Exit(2)
+				}
+				defer f.Close()
+				if err = pprof.WriteHeapProfile(f); err != nil {
+					log.Warningf("can't write %s: %s\n", path, err)
+				}
+				log.Infof("%s", path)
```
Comment (petermattis) [inline cli/start.go]: Yes, I meant to keep this as it logs whenever a memory profile is being written. Not sure of the utility though as you can simply see the files being created. Thoughts?

File: cli/start.go (hunk 1 above)
Comment (petermattis) [inline cli/start.go]: No. Copy-pasta.

**Comment by tamird** [comment]
:lgtm:

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> #### cli: add COCKROACH\_MEMPROF\_INTERVAL
> 
> The COCKROACH_MEMPROF_INTERVAL env var contains a duration which
> indicates how often memory profiles will be generated. Generated
> profiles are written to the logs directory with the prefix "memprof."
> and a date/time suffix. Periodic memory profiling is only initiated for
> the `start` command.
> 
> See #7018.
> 
> 
> 
</details>


---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[cli/start.go, line 128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuSRsHnix-kFM5Umq:-KJLuSRsHnix-kFM5Umr:287690729) ([raw file](https://github.com/cockroachdb/cockroach/blob/1950ebe1fe510936c78408f6795b079049878f68/cli/start.go#L128)):*
> ```Go
> 
> func initMemProfile(dir string) {
> 	memProfileInterval := envutil.EnvOrDefaultDuration("memprof_interval", -1)
> ```

nit: I think i'd introduce a `envutil.EnvDuration` method that returns `(time.Duration, bool)` for this.

---

*[cli/start.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLpTX1aHZgOKilB4hV-r1-157:-KJLuzzoqVpXw68kE1B8:-924208132) ([raw file](https://github.com/cockroachdb/cockroach/blob/1950ebe1fe510936c78408f6795b079049878f68/cli/start.go#L157)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, I meant to keep this as it logs whenever a memory profile is being written. Not sure of the utility though as you can simply see the files being created. Thoughts?
</details>
I think it's fine to leave, but I'd add some context rather than just logging a path.

---

*[cli/start.go, line 141 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLvGOFOusbGGio9naf:-KJLvGOFOusbGGio9nag:1564041360) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L141)):*
> ```Go
> 
> 	go func() {
> 		t := time.NewTicker(memProfileInterval)
> ```

defer t.Stop()

---

*[cli/start.go, line 150 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuvaMd6UHaE0OJR15:-KJLuvaMd6UHaE0OJR16:-232833601) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L150)):*
> ```Go
> 				f, err := os.Create(path)
> 				if err != nil {
> 					log.Warningf("%s", err)
> ```

nit: add some context to this

---

*[cli/start.go, line 155 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuxvpTOKLXzV5O0Z2:-KJLuxvpTOKLXzV5O0Z3:1255742174) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L155)):*
> ```Go
> 				defer f.Close()
> 				if err = pprof.WriteHeapProfile(f); err != nil {
> 					log.Warningf("can't write %s: %s\n", path, err)
> ```

no need for the newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-:-KJLvK3cqSrdFmisrAxo:1417088721)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[cli/start.go, line 128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuSRsHnix-kFM5Umq:-KJM7W55h4iHv-zo9CvE:971055177) ([raw file](https://github.com/cockroachdb/cockroach/blob/1950ebe1fe510936c78408f6795b079049878f68/cli/start.go#L128)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: I think i'd introduce a `envutil.EnvDuration` method that returns `(time.Duration, bool)` for this.
</details>
Seems a bit overkill. I'm fine with any negative duration to disable profiling.

---

*[cli/start.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLpTX1aHZgOKilB4hV-r1-157:-KJM7NbKeMLca4-RTYAi:-522027678) ([raw file](https://github.com/cockroachdb/cockroach/blob/1950ebe1fe510936c78408f6795b079049878f68/cli/start.go#L157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think it's fine to leave, but I'd add some context rather than just logging a path.
</details>
What additional context do you want? The path is pretty clearly a memory profile given the `memprof.` in the name.

---

*[cli/start.go, line 141 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLvGOFOusbGGio9naf:-KJM77bGfrZBLw4MOh4T:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L141)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> defer t.Stop()
</details>
Done.

---

*[cli/start.go, line 150 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuvaMd6UHaE0OJR15:-KJM7G0ptA75HxHWkCT3:1911976579) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L150)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: add some context to this
</details>
What additional context do you want? Currently this spits out messages like:

```
W160603 12:38:03.218271 cli/start.go:150  open cockroach-data/logs/blah/memprof.2016-06-03T12_38_03.218: no such file or directory
```

---

*[cli/start.go, line 155 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuxvpTOKLXzV5O0Z2:-KJM7Ldkd6C7eNhpHRr-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L155)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> no need for the newline
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cli/start.go, line 128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuSRsHnix-kFM5Umq:-KJM9jpg6gZNk7gNT2ln:160383661) ([raw file](https://github.com/cockroachdb/cockroach/blob/1950ebe1fe510936c78408f6795b079049878f68/cli/start.go#L128)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Seems a bit overkill. I'm fine with any negative duration to disable profiling.
</details>
OK.

---

*[cli/start.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLpTX1aHZgOKilB4hV-r1-157:-KJM9yAb08Zby1jA8twK:-1920233629) ([raw file](https://github.com/cockroachdb/cockroach/blob/1950ebe1fe510936c78408f6795b079049878f68/cli/start.go#L157)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What additional context do you want? The path is pretty clearly a memory profile given the `memprof.` in the name.
</details>
Yes, but it's not clear why it's being logged. In other words, what triggered this log line being printed?

---

*[cli/start.go, line 150 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuvaMd6UHaE0OJR15:-KJM9oIOGsGkR9R3L3FV:948689058) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L150)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What additional context do you want? Currently this spits out messages like:
> 
> ```
> W160603 12:38:03.218271 cli/start.go:150  open cockroach-data/logs/blah/memprof.2016-06-03T12_38_03.218: no such file or directory
> ```
</details>
Perhaps mention that no further attempts to profile memory will be made?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cli/start.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLpTX1aHZgOKilB4hV-r1-157:-KJMAaveXF0e0kHfkpSR:-172246595) ([raw file](https://github.com/cockroachdb/cockroach/blob/1950ebe1fe510936c78408f6795b079049878f68/cli/start.go#L157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yes, but it's not clear why it's being logged. In other words, what triggered this log line being printed?
</details>
I'm just going to remove this. It was mainly for debugging and it is easy enough to see if memprof files are being created.

---

*[cli/start.go, line 150 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuvaMd6UHaE0OJR15:-KJMAUdW8bq22mcRomEH:-346680321) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L150)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Perhaps mention that no further attempts to profile memory will be made?
</details>
We're in a function here. We'll try again after the interval. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[cli/start.go, line 150 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuvaMd6UHaE0OJR15:-KJMDqkwBb15Aci35jrv:82424682) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L150)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We're in a function here. We'll try again after the interval. 
</details>
Ah, missed that. I still feel like I'd want context, but OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cli/start.go, line 150 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027#-KJLuvaMd6UHaE0OJR15:-KJMJGaz1cyIHbpiKZsZ:252288101) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e44b19ec2cc98e2289fd8c622fbf1126b3a0bc5/cli/start.go#L150)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Ah, missed that. I still feel like I'd want context, but OK.
</details>
I doubt there will be any confusion in practice and this code path is likely never going to be taken because being unable to write to the log directory will cause other problems.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7027)*

**Pull request merged**